design/general-progress-monitoring.md

```diff
@@ -0,0 +1,515 @@
+                                                // total size to transfer (may be less than the size of a volume if
+                                                // performing an incremental) and itemsCompleted is the number of bytes
+                                                // transferred.  On successful completion, itemsCompleted and itemsToComplete
+                                                // should be the same
+        started, updated time.Time              // When the upload was started and when the last update was seen.  Not all
+                                                // systems retain when the upload was begun, return Time 0 (time.Unix(0, 0))
+                                                // if unknown.
+    }
+
+### VolumeSnapshotter changes
+
+Two new methods will be added to the VolumeSnapshotter interface:
+
+    Progress(snapshotID string) (OperationProgress, error)
+    Cancel(snapshotID string) (OperationProgress, error)
+
+Open question: Does VolumeSnapshotter need Cancel, or is that only needed for BIA/RIA?
+
+Progress will report the current status of a snapshot upload.  This should be callable at
+any time after the snapshot has been taken.  In the event a plug-in is restarted, if the operationID
+(snapshot ID) continues to be valid it should be possible to retrieve the progress.
+
+`error` is set if there is an issue retrieving progress.  If the snapshot is has encountered an
+error during the upload, the error should be returned in OperationProgress and error should be nil.
+
+### BackupItemAction and RestoreItemAction plug-in changes
+
+Currently CSI snapshots and the Velero Plug-in for vSphere are implemented as BackupItemAction
+plugins.  While the majority of BackupItemAction plugins do not take snapshots or upload data, this
+functionality is useful for any longstanding plugin operation managed by an external
+process/controller so we will modify BackupItemAction and RestoreItemAction to optionally return an
+operationID in addition to the modified item.
+
+Velero can attempt to cancel an operation by calling the Cancel API call on the BIA/RIA. The plugin
+can then take any appropriate action as needed. Cancel will be called on backup deletion, and
+possibly upon reaching timeouts, if timeout support is included in this feature. Cancel is not
+intended to be used to delete/remove the results of completed actions.
+
+_AsyncOperationsNotSupportedError_ should only be returned (by either Cancel or by Progress) if the
+Backup/RestoreItemAction plugin should not be handling the item.  If the Backup/RestoreItemAction
+plugin should handle the item but, for example, the item/snapshot ID cannot be found to report
+progress, Progress or Cancel will return an InvalidOperationIDError error rather than a populated
+OperationProgress struct. If the item action does not start an asynchronous operation, then
+operationID will be empty.
+
```

Comment:
Should we also clarify the expected return values if `Progress` or `Cancel` is called against a canceled operation?  Will it be an error or somehow reflected in the `OperationProgress`?

Reply:
Cancel just returns an error if something unexpected happens. Calling cancel again shouldn't force an error -- it's up to the plugin, but I'd expect a second Cancel call would be a no-op (and Velero shouldn't actually call it twice unless there's a bug).

For Progress, again, this shouldn't happen, and the actual behavior is up to the plugin, but I'd expect the plugin to do one of two things based on the overall design already identified:
1) if the canceled operation is no longer available (i.e. the CR associated with it has been deleted by its controller), then it should return a standard "not found" error just as it would with an operationID that's never been valid
2) If the canceled operation *is* still there, then I'd expect progress to do what it normally does -- if it's actually in progress (cancel didn't work/isn't supported), then report that. If there's an error state, then report that. If the plugin has information on completion stats, report that. I don't know that it's really a separate case that Velero should care about, since Velero will never call Progress or Cancel after calling Cancel on an operation.

Reply:
> since Velero will never call Progress or Cancel after calling Cancel on an operation.
So it means the Cancel is a sync operation right? i.e. if everything is fine it won't return until the operation is really canceled.  If this is the case, could you explicitly define that in this design？

Reply:
If everything is fine, Velero will never call cancel. The only time Velero will call Cancel is if we've hit the timeout and the operation is still ongoing. However, once we call cancel, since it's up to the plugin what to do with it (could be a no-op if the plugin can't cancel), once Cancel returns, velero moves on. Basically the Cancel API call is "best effort" -- Velero doesn't guarantee that the cancel was successful, since some plugins may not be able to cancel at all. Velero calls cancel and moves on. I think that was stated explicitly, but if it's not clear enough, let me know and I can try to reword that somewhere.

Reply:
@sseago 
Thanks for the reply. 

Since you are putting details like error types in this section, IMO, the behavior of `Cancel` func of a plugin should be be clarified.
I suggest we add more details in the paraphrase  `Velero can attempt to cancel an operation by calling the Cancel API call on the BIA/RIA....`.   We should clarify the contract between the velero and the plugin in terms of `Cancel` func.  For example, when this function returns `nil`, does it indicates (a) the operation is canceled successfully or (b)  the cancel request is handled but the result is unknown?

I presonally prefer option (a),  if it's not possible for the plugin to know if the operation is really canceled, it can return a `status unknown error` and velero may choose to ignore it.
If you think (b) is a better choice, I'd suggest we clarify in what case the `Cancel` func returns a non-nil value, or maybe it'd be better that the `Cancel` func does not have any values.  So the caller just call the func to attempt to cancel the operation, and it may or may not be successful.

Either way, I hope this can be clarified in this design, otherwise, the plugin developer will be confused when he/she is implementing the `Cancel` func.

Reply:
There is no expectation for cancel to return an error under normal operations. Since any plugin-implemented cancel request is likely to be async as well, there's no way to provide velero with reliable "this was successfully cancelled" information in the general case, since velero won't be checking back. The whole point of cancel from velero's point of view is "we hit the timeout, we're declaring this operation failed, and we're giving the plugin an opportunity to cancel on its end if it can" -- but velero is logging this as an error either way. This means the plugin doesn't need to tell velero anything. A plugin which doesn't support cancel will implement a no-op func -- returning nil for error. A cancel-supporting plugin will do whatever it needs to cancel and return nil for error (and it probably won't know at this point whether cancel was successful). An error will be returned if the plugin itself gets an unexpected error in cancellation.

Reply:
> A plugin which doesn't support cancel will implement a no-op func -- returning nil for error. A cancel-supporting plugin will do whatever it needs to cancel and return nil for error (and it probably won't know at this point whether cancel was successful). An error will be returned if the plugin itself gets an unexpected error in cancellation.

This sounds OK to me.  Let's make sure this is documented, in the .md file and in the inline comment of the go interface.